### DISCUSSION TITLE
Example Usage - handle tripples in a loop

### DISCUSSION THREAD

**Discussion by hoijui** [pchampin/sophia_rs#102] [Q&A] [answered]
Starting with the code at https://docs.rs/sophia_api/0.6.2/sophia_api/graph/trait.Graph.html#examples ,
I wanted to fill the loop with code, but failed using `triple`, due to borrowing/life-time issues.
One trial can be seen in this minimal project, where I want to store the parts of the triple in `vec`'s, for using them in some way, later:
https://github.com/hoijui/sophia-q1

I also had problems using the data-structure.. I imagine I have to get the `Rc<str>` somehow, but I do not know how, and if that is right.

Could anyone give a heads up?
I would also be happy to know how to get there using the compiler errors.

**Comment by pchampin** [comment, marked as answer]
Disclaimer: I understand that the source of your problem is the lack of documentation in Sophia, so in the end, I am the one to blame.

That being said, it is hard to help you solve this particular problem as it is. The code you are pointing generates a lot of warnings, and the error reported by the compiler is clearly not the one that causes you problem (calling the fake method `bla()`).

For me to help you efficiently, could you please try to build a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) so that we could focus on one problem at a time?

**Reply by pchampin** [reply]
That being said, I just created a pull-request on your example repo with a running version of (what I understand of) you are trying to do, and some explanations of the rationale...

Ownership and lifetimes are hard to get right for beginners (and, I must says, not-quite-beginners as well). But overcoming those difficulties is very rewarding, and these notions are what makes Rust great. Good luck with your learning.

PS: if you have Rust questions that are not directly related to Sophia, don't hesitate to go to StackOverflow. The Rust community there is super welcoming and incredibly helpful.

**Comment by hoijui** [comment]
Perfect!
Thank you @pchampin !
your [pull request](https://github.com/hoijui/sophia-q1/pull/1) is perfect! :-)

... and yes.. I was unsure if it makes sense to treat this as a problem I have with sophia, or rather one I have with rust. I can only say that I had less problems using these crates: rdf, kiss3d, petgraph

Thanks a lot, again!
I felt I was stuck at this with learning rust, and would not have continued without your help. I could not even have framed it as a general rust question (to put on SO, for example), as I could not pin it down to something specific.
ride on!

**Reply by pchampin** [reply]
One possible reason why it is harder with Sophia is that Sophia is intended to be very generic. Most of the methods are defined in *traits*, and how to actually use them depends on which specific *implementation* of these traits you are using.... best
